src/mesh/voxel_cloud.rs

```diff
@@ -0,0 +1,456 @@
+/// therefore its beginning as well as the voxel positions are defined in the
+/// voxel-space coordinates. The voxel space starts at the cartesian space
+/// origin with voxel coordinates 0, 0, 0. Voxel clouds with the same voxel size
+/// are compatible and collateral operations be performed on them.
+#[derive(Debug, Clone)]
+pub struct VoxelCloud {
+    block_start: Point3<i32>,
+    block_dimensions: Vector3<u32>,
+    voxel_dimensions: Vector3<f32>,
+    // FIXME: @Optimization Change this to a bit vector
+    voxel_map: Vec<bool>,
+}
+
+impl VoxelCloud {
+    /// Define a new empty block of voxel space, which begins at
+    /// `block_start`(in discrete voxel units), has dimensions
+    /// `block_dimensions` (in discrete voxel units) and contains voxels sized
+    /// `voxel_dimensions` (in model space units).
+    pub fn new(
+        block_start: &Point3<i32>,
+        block_dimensions: &Vector3<u32>,
+        voxel_dimensions: &Vector3<f32>,
+    ) -> Self {
+        assert!(
+            block_dimensions.x > 0 && block_dimensions.y > 0 && block_dimensions.z > 0,
+            "One or more block dimensions are 0"
+        );
+        assert!(
+            voxel_dimensions.x > 0.0 && voxel_dimensions.y > 0.0 && voxel_dimensions.z > 0.0,
+            "One or more voxel dimensions are 0.0"
+        );
+        let map_length = block_dimensions.x * block_dimensions.y * block_dimensions.z;
+        VoxelCloud {
+            block_start: *block_start,
+            block_dimensions: *block_dimensions,
+            voxel_dimensions: *voxel_dimensions,
+            voxel_map: vec![false; cast_usize(map_length)],
+        }
+    }
+
+    /// Calculate a voxel cloud from an existing mesh.
+    pub fn from_mesh(mesh: &Mesh, voxel_dimensions: &Vector3<f32>) -> Self {
+        // Determine the needed block of voxel space.
+        let b_box = BoundingBox::from_meshes(iter::once(mesh));
+
+        let min_point = &b_box.minimum_point();
+        let max_point = &b_box.maximum_point();
+        let min_x_index = (min_point.x.min(max_point.x) / voxel_dimensions.x).floor();
+        let max_x_index = (min_point.x.max(max_point.x) / voxel_dimensions.x).ceil();
+        let min_y_index = (min_point.y.min(max_point.y) / voxel_dimensions.y).floor();
+        let max_y_index = (min_point.y.max(max_point.y) / voxel_dimensions.y).ceil();
+        let min_z_index = (min_point.z.min(max_point.z) / voxel_dimensions.z).floor();
+        let max_z_index = (min_point.z.max(max_point.z) / voxel_dimensions.z).ceil();
+
+        let block_start = Point3::new(
+            min_x_index as i32 - 1,
+            min_y_index as i32 - 1,
+            min_z_index as i32 - 1,
+        );
+        let block_dimensions = Vector3::new(
+            (max_x_index - min_x_index + 2.0) as u32,
+            (max_y_index - min_y_index + 2.0) as u32,
+            (max_z_index - min_z_index + 2.0) as u32,
+        );
```

Comment:
better safe than sorry. but you are right, `floor` and `ceil` should suffice.

Reply:
actually, it panicks without a positive padding and makes holes without a negative one. perhaps i should investigate more but i suspect the rounding in cartesian to voxel coords calculations might put a model-space coord out of voxel block bounds. that explains just the panic, though.

Reply:
catched both bugs:

1. when converting the voxel coordinates into boolean vector index, it's not enough to check if the resulting index is within bounds but rather it's necessary to check each voxel XYZ to be within theyr respecitve bounds
2. the block dimensions are `max_index + 1`, because the index is always 1 lower than the length

Reply:
Good job! Yeah, I totally forgot that indexing at `[3, 3, 3]` in a voxel space `[2, 2, 10]` ends up within bounds of the array, but not the voxel space.